Jacobi007 jk add include paths

ABaldwinHunter opened the pull request:
resolved merge conflicts to incorporate changes from https://github.com/jacobi007/codeclimate-fixme/tree/jk-add-include-paths

- Makes Fixme engine accept `include_paths`

@codeclimat/review

GordonDiggs commented:
I believe this repo has automated tests now, can you add some?

GordonDiggs commented on index.js:
```diff
@@ -65,39 +65,67 @@ var findFixmes = function(file){
-  !fs.lstatSync(fp).isDirectory() &&
-  (excludeExtensions.indexOf(path.extname(fp)) < 0)
+// Returns an array of unique array values not included in the other provided array
+var diff = function(a1, a2) {
```
What do you think of putting this in its own module and requiring it at the top of the file?

GordonDiggs replied:
Would be something like this (I think):

```javascript
// diff.js
module.exports = function(a1, a2) {
...
};

// index.js
var diff = require("diff");
````

ABaldwinHunter replied:
sounds good! Adding tests too ...

GordonDiggs commented on index.js:
```diff
@@ -65,39 +65,67 @@ var findFixmes = function(file){
+// Filters the directory paths out
+var filterFiles = function(files) {
+  return files.filter(function(file) {
+    return !fs.lstatSync(file).isDirectory();
```
Is there an `isFile` we can use to avoid symlinks? Or am I wrong about our policy on symlinks?

ABaldwinHunter commented:
@codeclimate/review @GordonDiggs 

Ready for another look!

Main changes:
- Makes fixme engine use include paths
- adds testing suite with mocha and chai, integrated with circle. One real spec, a few pending
- refactors the Fixme code from one large file into a couple of smaller files. This change makes the code more readable and makes it easier to package the tooling in future.

The circle build runs tests! https://circleci.com/gh/codeclimate/codeclimate-fixme/35

ABaldwinHunter commented:
It also does some renaming of files, and updates the .codeclimate.yml to use eslint.

*Adding tests feels important, but it felt like a large enough change to add the main suite and structure. We can make tests more robust in separate PR.

ABaldwinHunter commented:
P.S. I can also squash commit history if preferred.

GordonDiggs commented on bin/fixme:
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env node
 
-var FixMe = require('../index');
+var FixMe = require('../lib/fixMe');
```
I think generally all filenames should be underscored (but the variable name looks good)

GordonDiggs commented on circle.yml:
```diff
@@ -8,9 +8,16 @@ machine:
+dependencies:
   override:
     - docker build -t=$PRIVATE_REGISTRY/$CIRCLE_PROJECT_REPONAME:b$CIRCLE_BUILD_NUM .
+    - npm install mocha
```
These should be listed in the package.json file and you should just run `npm install`. I can help you with that tomorrow!

GordonDiggs commented on lib/fileBuilder.js:
```diff
@@ -0,0 +1,42 @@
+  filterFiles: filterFiles
+};
+
+module.exports = fileBuilder;
```
To pair on: make this more idiomatic by inlining the object

ABaldwinHunter replied:
Thanks for the feedback! Working on this and other suggestions. Hit me up if you'd like to pair later :)

GordonDiggs commented on .codeclimate.yml:
```diff
@@ -1,8 +1,14 @@
 engines:
   fixme:
+    enabled: false
```
should this be true?

ABaldwinHunter replied:
It could be! 

Enabling seemed to require excluding the main files of the project (because they have the name `Fixme` peppered throughout), so ESLint seemed more valuable to me.

I can revert and/or think about a way to avoid false positives from Fixme

ABaldwinHunter replied:
p.s. I think engine-specific exclude_paths are in the works. 

This seems like an excellent use case!

GordonDiggs replied:
oh hah yeah let's leave it off for now

GordonDiggs commented on .codeclimate.yml:
```diff
@@ -1,8 +1,14 @@
   fixme:
+    enabled: false
+  eslint:
     enabled: true
```
Can you also add our default `.eslintrc` file?

ABaldwinHunter replied:
sure sounds good!

GordonDiggs commented:
Minor CC nits but LGTM! I think it might be good to squash the commits before you merge.

ABaldwinHunter commented:
@GordonDiggs thanks for the feedback! I added .eslintrc.

I also added a few things to our `package.json`. Could you take another look when you have a moment?

new stuff
```
 "main": "./lib/fix-me.js",
  },
  "files": [
    "LICENSE",
    "README.md",
    "bin",
    "lib"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/codeclimate/codeclimate-fixme"
  },
  "keywords": [
    "fixme",
    "codeclimate"
  ],
  "engines": {
    "node": ">=0.10"
  },
  "license": "MIT"
```
If those changes look okay, I'll go ahead and squash commits.

GordonDiggs commented on .eslintrc:
```diff
@@ -0,0 +1,8 @@
+    "node": true
+  },
+  "rules": {
+    "strict": 0
```
Our default file [is much bigger](https://github.com/codeclimate/codeclimate/blob/master/config/eslint/.eslintrc). What made you go with this?

ABaldwinHunter replied:
Oh cool. Updated.

I originally tried to mimic https://github.com/codeclimate/codeclimate-eslint/blob/master/.eslintrc

ABaldwinHunter replied:
I can update that one to match after

GordonDiggs replied:
Ah yeah that one needs to be changed

GordonDiggs commented on package.json:
```diff
@@ -0,0 +1,34 @@
+  ],
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/codeclimate/codeclimate-fixme"
```
This should be the git url, so `https://github.com/codeclimate/codeclimate-fixme.git`

ABaldwinHunter replied:
Thanks! Good catch

GordonDiggs commented:
LGTM
